### PR TITLE
Data extensibility for `Block` and `BlockBody`

### PR DESCRIPTION
**Pull request by rafal-ch**
Merge data extensibility PoC (`Block` and `BlockBody`) into `feat-2.0`.

Fixes: #4240

### CODE REVIEW

**Comment by rafal-ch** [comment]
`do-not-merge`, still need to figure out the NCTL tests

**Update:** the `upgrade_scenario_01.sh` (which is currently the only one executed by `nctl_upgrade.sh` is not affected by the changes to block RPC schema and doesn't require client updates. There is a "[temporary hack](https://github.com/rafal-ch/casper-node/commit/72ce59cbfeb8179bc375c2f0be7ec1d8048b84ef)" prepared on the side branch which we can use to move forward, should the refactoring of `MetaBlock` require more work than expected.

With the mentioned commit in place, the NCTL upgrade scenario passes successfully on the `feat-data-extensibility` branch.

cc: @alsrdn

File: node/src/components/gossiper/provider_impls/block_provider.rs [hunk 1]
```diff
@@ -36,11 +36,15 @@ impl ItemProvider<Block> for Gossiper<{ Block::ID_IS_COMPLETE_ITEM }, Block> {
...
         item_id: BlockHash,
-    ) -> Option<Box<Block>> {
+    ) -> Option<Box<BlockV2>> {
         // TODO: Make `get_block_from_storage` return a boxed block instead of boxing here.
-        effect_builder
-            .get_block_from_storage(item_id)
-            .await
-            .map(Box::new)
+        if let Some(block) = effect_builder.get_block_from_storage(item_id).await {
+            match block {
```
Comment (alsrdn) [inline node/src/components/gossiper/provider_impls/block_provider.rs]: nit: we can skip matching here `std::convert::TryInto::<BlockV2>::try_into(block).ok().map(|block| Box::new(block));`

File: types/src/block.rs [hunk 2]
```diff
@@ -3,54 +3,57 @@ mod block_hash;
...
+#[cfg(any(feature = "testing", test))]
+pub mod test_block_builder;
 
-#[cfg(any(all(feature = "std", feature = "testing"), test))]
-use alloc::collections::BTreeMap;
 use alloc::{boxed::Box, vec::Vec};
 use core::fmt::{self, Display, Formatter};
-#[cfg(any(all(feature = "std", feature = "testing"), test))]
-use core::iter;
+#[cfg(feature = "once_cell")]
```
Comment (alsrdn) [inline types/src/block.rs]: Shouldn't this be `#[cfg(feature = "json-schema")]` instead since we use `Lazy` only for defining `static BLOCK` below?

**Review by alsrdn — APPROVED** [review]
One small I think with a feature guarded use, but otherwise LGTM.

File: node/src/components/gossiper/provider_impls/block_provider.rs (hunk 1 above)
Comment (rafal-ch) [reply node/src/components/gossiper/provider_impls/block_provider.rs]: I used a little more concise form: `block.try_into().ok().map(Box::new)`
Updated here: d0b7db9e4d3a34794f455a209655d4869ee36077

File: types/src/block.rs (hunk 2 above)
Comment (rafal-ch) [reply types/src/block.rs]: Yes, good catch. Fixed in 60d1b481e6b1ea24a92343c49ce6b724cb15f6b9 and 06fff783c44ae7a795387e3cc60c63bee9784c93 along with the update to `json_block_with_signatures.rs`.

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**